openstack/02a-MAClow/IEEE802154_security.c

```diff
@@ -0,0 +1,905 @@
+
+      // Keep a pointer to where the ASN will be
+      // Note: the actual value of the current ASN will be written by the
+      //    IEEE802.15.4e when transmitting
+      msg->l2_ASNpayload = msg->payload;
+   }
+
+   //security control field
+   packetfunctions_reserveHeaderSize(msg, sizeof(uint8_t));
+ 
+   temp8b = 0;
+   temp8b |= msg->l2_securityLevel << ASH_SCF_SECURITY_LEVEL;//3b
+   temp8b |= msg->l2_keyIdMode << ASH_SCF_KEY_IDENTIFIER_MODE;//2b
+   temp8b |= frameCounterSuppression << ASH_SCF_FRAME_CNT_MODE; //1b
+
+   if (ieee802154_security_vars.m_macFrameCounterMode == 0x04){
+      temp8b |= 0 << ASH_SCF_FRAME_CNT_SIZE; //1b
+   } else{
+      temp8b |= 1 << ASH_SCF_FRAME_CNT_SIZE; //1b
+   }
+
+   temp8b |= 0 << 1;//1b reserved
+   *((uint8_t*)(msg->payload)) = temp8b;
+}
+
+/**
+\brief Key searching and encryption/authentication operations.
+*/
+owerror_t IEEE802154_security_outgoingFrameSecurity(OpenQueueEntry_t*   msg){
+   bool frameCounterSuppression;
+   m_keyDescriptor* keyDescriptor;
+   uint8_t i;
+   uint8_t j;
+   uint8_t nonce[13];
+   uint8_t key[16];
+   owerror_t outStatus;
+   uint8_t* a;
+   uint8_t len_a;
+   uint8_t* m;
+   uint8_t len_m;
+
+   //the frame counter is carried in the frame, otherwise 1;
+   frameCounterSuppression = 1;
+
+   //search for a key
+   keyDescriptor = IEEE802154_security_keyDescriptorLookup(msg->l2_keyIdMode,
+                                                          &msg->l2_keySource,
+                                                          msg->l2_keyIndex,
+                                                          &msg->l2_keySource,
+                                                          (idmanager_getMyID(ADDR_PANID)),
+                                                          msg->l2_frameType);
+
+   if (keyDescriptor==NULL){
+      openserial_printError(COMPONENT_SECURITY,ERR_SECURITY,
+                          (errorparameter_t)msg->l2_frameType,
+                          (errorparameter_t)0);
+      return E_FAIL;
+   }
+
+   for (j=0;j<16;j++){
+      key[j] = keyDescriptor->key[j];
+   }
+
+   uint8_t vectASN[5];
+   macFrameCounter_t l2_frameCounter;
+   ieee154e_getAsn(vectASN);//gets asn from mac layer.
+   if (frameCounterSuppression == 0){//the frame Counter is carried in the frame
+      //save the frame counter of the current frame
+      l2_frameCounter.bytes0and1 = vectASN[0]+256*vectASN[1];
+      l2_frameCounter.bytes2and3 = vectASN[2]+256*vectASN[3];
+      l2_frameCounter.byte4 = vectASN[4];
+
+      IEEE802154_security_getFrameCounter(l2_frameCounter,
+                                         msg->l2_ASNpayload);
+   } //otherwise the frame counter is not in the frame
+
+   //nonce creation
+   memset(&nonce[0], 0, 13);
+   switch (msg->l2_keyIdMode){
+      case 0:
+      case 1:
+         for (i=0; i<8; i++){
+            nonce[i] = ieee802154_security_vars.m_macDefaultKeySource.addr_64b[i];
```

Comment:
@ssciancalepore, 15.4 nonce should be created from extended/short address independent of keyIdMode